databuilder/dsl.py

```diff
@@ -302,18 +345,23 @@ def apply_to(self, table: BaseTable) -> BaseTable:
 
 
 @dataclass
-class Column:
+class Column(Generic[S]):
```

Comment:
I'm struggling a bit with understanding what this and the `TypeVar` does.  Does this make Column a PatientSeries in some way?

Reply:
I'm going to try and explain this one (@benbc please correct me if I'm wrong!)…

This line makes `Column` a generic Type with the placeholder/typevar `S` to denote the concrete type we pass around.  The TypeVar's `bound` kwarg says that any concrete type used must be a PatientSeries or a subclass of it.

Each Column then declares which Series class it is related to, eg `class IdColumn(Column[IdSeries]):` and sets the appropriate `series_type` instance variable with the relevant Series class in it's `__init__()`.

We've then set up `Column.series_type: type[S]` so that `PatientFrame.select_column()` can return the appropriate Series classed based on the column asked for.  The type system knows how to tie this all together because we're using that same `S` type var in the `select_column()` definition as we are here.

Reply:
Thanks!

Reply:
Yes, I think that's right.

The most important thing is that each column now carries with it information about what kind of series will be created if that column is selected. That information is present both in the runtime object model (the `series_type` field) and in the type system (the type var `S`).